### PR TITLE
Add command for displaying image streams

### PR DESCRIPTION
**Pull request by NicoBiernat**
Some things that could be improved / changed:
- use futures instead of polling the stream and key events
- integrate this into the "cat" command
- add a border and title around the display

### CODE REVIEW

File: src/cmd/mod.rs [hunk 1]
```diff
@@ -31,4 +31,5 @@ cmd_mods! {
     touch,
     tree,
     uln,
+    display,
```
Comment (fwcd) [inline src/cmd/mod.rs]: Nit: These are ordered alphabetically

File: Cargo.toml [hunk 2]
```diff
@@ -14,8 +14,11 @@ async-recursion = "1.1.0"
 clap = { version = "4.5", features = ["derive", "env"] }
 colored = "2.1.0"
 dotenvy = "0.15"
+futures = "0.3.30"
+futures-util = "0.3.30"
```
Comment (fwcd) [inline Cargo.toml]: IIRC `futures` is already an umbrella crate that includes `futures-util`.

```suggestion
```

File: src/cmd/display.rs [hunk 3]
```diff
@@ -0,0 +1,108 @@
+use crate::{context::Context, path::VirtualPathBuf};
+use anyhow::Result;
+use clap::{command, Parser};
+use futures::poll;
+use futures_util::stream::StreamExt;
```
Comment (fwcd) [inline src/cmd/display.rs]:
```suggestion
use futures::{poll, StreamExt};
```

File: src/cmd/display.rs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    let path = ctx.cwd.join(args.path);
+
+    let mut stream = ctx.lh.stream::<Value, Model>(&path.as_lh_vec(), Value::Nil).await?;
+
+    enable_raw_mode()?;
+    stdout().execute(EnterAlternateScreen)?;
+
+    let mut terminal = Terminal::new(CrosstermBackend::new(stdout()))?;
+    loop {
+        if let Poll::Ready(Some(Ok(msg))) = poll!(stream.next()) {
```
Comment (fwcd) [inline src/cmd/display.rs]: This is an interesting way to go about it. Feels a bit unusual to poll a future manually in an async context (this is only very rarely done in high-level async code), but I'm not sure whether there is a better alternative.

Another option that I could think of, off top of my head, would be to spawn a separate task that runs a receive loop on the stream and forwards the messages into a channel/queue that we poll here. Something like (pseudocode, no idea whether this compiles):

```rust
let (sender, receiver) = channel();

tokio::spawn(async move {
  while let Some(msg) = stream.next().await {
    sender.send(msg);
  }
});

loop {
  if let Ok(msg) = receiver.try_recv() {
    ...
  }
}
```

Perhaps we would have to use either some form of signalling mechanism or store the handle to the spawned task to cancel it (the stream) once the user exits the command.

**Review by fwcd — COMMENTED** [review]
Thanks, this looks pretty cool and I look forward to having a way of viewing streamed frames graphically in real-time! I have left some, mostly technical notes, below.

One remark:

> integrate this into the "cat" command

That's an interesting idea, but I would probably keep it in a separate command, just to keep things simple and because there doesn't seem to be an obvious Unix analogon to this functionality.

File: src/cmd/display.rs (hunk 4 above)
Comment (NicoBiernat) [reply src/cmd/display.rs]: The problem is that we need to wait for either a new image from the stream or a keyboard event from the ratatui library in order to exit the display. The current solution is polling both in a busy loop that is only limited by the keyboard event polling timeout.

Maybe we could spawn two tasks that poll the images and events and forwards them to a single stream that is then awaited in the main task.
Or is there something like the "select" keyword in Go, which allows you to receive from multiple channels simultaneously and process the one that has a value first?
I haven't worked with tokio and channels in Rust before, so I tried things until they worked :D

Comment (fwcd) [reply src/cmd/display.rs]: There's a `select!` macro: https://rust-lang.github.io/async-book/06_multiple_futures/03_select.html

I haven't used it yet, but might be worth a try.

**Comment by NicoBiernat** [comment]
The application still hangs after a while for unknown reasons...

**Comment by fwcd** [comment]
Thanks again!

I have merged the changes from the main branch, which updates `lighthouse-client` to 3.4.0. In my tests, it seems to run pretty reliably, so hopefully these issues are fixed. If not, feel free to open another issue.

The canvas now uses a custom `Shape` implementation that makes sure to render the display at an integer scale to make sure that no rows or columns are skipped. Finally, the display now also forwards all key strokes other than `q` (which quits the command) as input events to the server, enabling games etc.

**Pull request merged**